modules/gapi/test/common/gapi_tests_common.hpp

```diff
@@ -117,6 +142,142 @@ using compare_f = std::function<bool(const cv::Mat &a, const cv::Mat &b)>;
+        constexpr int specific_params_size = std::tuple_size<specific_params_t>::value;
+        copyValues(params, m_common,
+            typename detail::MkRange<0, common_params_size>::type());
+        copyValues(params, m_specific,
+            typename detail::MkRange<common_params_size, specific_params_size>::type());
+    }
+public:
+    Params() = default;
+    Params(const params_t& params)
+    {
+        init(params);
+    }
+    Params& operator=(const params_t& params)
+    {
+        init(params);
+        return *this;
+    }
+
+    const common_params_t& commonParams() const
+    {
+        return m_common;
+    }
+
+    const specific_params_t& specificParams() const
+    {
+        return m_specific;
+    }
+};
+
+template<>
+class Params<>
+{
+public:
+    using common_params_t = std::tuple<int, cv::Size, int, bool, cv::GCompileArgs>;
+    using specific_params_t = std::tuple<>;
+    using params_t = common_params_t;
+private:
+    params_t m_all;
+public:
+    Params() = default;
+    Params(const params_t& params) : m_all(params) {}
+    Params& operator=(const params_t& params)
+    {
+        m_all = params;
+        return *this;
+    }
+
+    const common_params_t& commonParams() const
+    {
+        return m_all;
+    }
+};
+
+template<typename ...SpecificParams>
+class TestWithParamBase : public TestFunctional,
+    public TestWithParam<typename Params<SpecificParams...>::params_t>
+{
+    Params<SpecificParams...> m_params;
+
+    void init(TestWithParamBase* instance)
+    {
+        using TestWithParamClass =
+            TestWithParam<typename Params<SpecificParams...>::params_t>;
+        instance->m_params = instance->TestWithParamClass::GetParam();
+        std::tie(instance->type, instance->sz, instance->dtype, instance->createOutputMatrices,
+            instance->compile_args) = instance->m_params.commonParams();
+        if (instance->dtype == SAME_TYPE)
+        {
+            instance->dtype = instance->type;
+        }
+        initRand(instance->type, instance->sz, instance->dtype, instance->createOutputMatrices,
+            instance->distribution);
+    }
+public:
+    using common_params_t = typename Params<SpecificParams...>::common_params_t;
+    using specific_params_t = typename Params<SpecificParams...>::specific_params_t;
+
+    MatType type = -1;
+    cv::Size sz = {};
+    MatType dtype = -1;
+    bool createOutputMatrices = false;
+    cv::GCompileArgs compile_args = {};
+    int distribution = -1;
+
+    TestWithParamBase(int _distributionType = cv::RNG::NORMAL) :
+        distribution(_distributionType)
+    {
+        init(this);
+    }
+
+    const Params<SpecificParams...>& GetParam() const
+    {
+        return m_params;
+    }
+};
+
+#define USE_UNIFORM_INIT(class_name) \
+    class_name() : TestWithParamBase(cv::RNG::UNIFORM) {}
+
+#define USE_NORMAL_INIT(class_name) \
+    class_name() : TestWithParamBase(cv::RNG::NORMAL) {}
```

Comment:
hmm. If this is all about specifying some enum, y not
1. Use CRTP and define some new macro to define tests (macro would pass the derived type to `TestWithParamBase` as template argument. Something like
```c++
#define GAPI_TEST(Class, Signature)   struct Class: public TestWithParamBase<Class, std::tuple Signature >
```
2. These macros become
```
#define USE_INTERPOLATION(X) constexpr int INTERPOLATION = X;
```
3. In the base class you just refer to interpolation via CRTP (or to a default value if interpolation is not defined)

Reply:
I thought about this. Probably, it would've been even simpler if `TestWithParamBase` accepted `int` as a first template argument:
```cpp
struct Class: public TestWithParamBase<INTERPOLATION, std::tuple Signature >
```
Anyway, I propose to leave it out for now and add separately (in the following PR)

Reply:
ok!